File: Bank.py
import json
import os
from datetime import datetime

DATA_FILE = "bank_data.json"

class Transaction:
    @staticmethod
    def create(amount: float, t_type: str, details: str = "") -> dict:
        return {
            "timestamp": datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            "amount": amount,
            "type": t_type,
            "details": details
        }

    @staticmethod
    def format(t: dict) -> str:
        return f"[{t['timestamp']}] {t['type'].ljust(20)} | {t['amount']:>10.2f} PLN | {t['details']}"


class Bank:
    def __init__(self):
        self.accounts = self._load_data()

    def _load_data(self) -> dict:
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r', encoding='utf-8') as f:
                    return json.load(f)
            except Exception:
                return {}
        return {}

    def _save_data(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump(self.accounts, f, indent=4, ensure_ascii=False)

    def create_account(self, acc_number: str, owner: str, pin: str):
        if acc_number in self.accounts:
            return False, "Konto o tym numerze już istnieje."
        
        self.accounts[acc_number] = {
            "owner": owner,
            "pin": pin,
            "balance": 0.0,
            "history": []
        }
        self._save_data()
        return True, "Konto zostało pomyślnie utworzone."

    def authenticate(self, acc_number: str, pin: str):
        account = self.accounts.get(acc_number)
        if account and account["pin"] == pin:
            return account
        return None

    def deposit(self, acc_number: str, amount: float):
        if amount <= 0:
            return False, "Kwota musi być większa niż zero."
        
        self.accounts[acc_number]["balance"] += amount
        self.accounts[acc_number]["history"].append(
            Transaction.create(amount, "Wpłata", "Wpłata własna")
        )
        self._save_data()
        return True, "Wpłata zakończona sukcesem."

    def withdraw(self, acc_number: str, amount: float):
        if amount <= 0:
            return False, "Kwota musi być większa niż zero."
        
        if self.accounts[acc_number]["balance"] < amount:
            return False, "Brak wystarczających środków."

        self.accounts[acc_number]["balance"] -= amount
        self.accounts[acc_number]["history"].append(
            Transaction.create(-amount, "Wypłata", "Wypłata w gotówce")
        )
        self._save_data()
        return True, "Wypłata zakończona sukcesem."

    def transfer(self, from_acc: str, to_acc: str, amount: float):
        if amount <= 0:
            return False, "Kwota przelewu musi być większa niż zero."
        if from_acc == to_acc:
            return False, "Nie można wykonać przelewu na to samo konto."
        if to_acc not in self.accounts:
            return False, "Konto docelowe nie istnieje."
        if self.accounts[from_acc]["balance"] < amount:
            return False, "Brak wystarczających środków na koncie."

        # Atomowa operacja w słowniku
        self.accounts[from_acc]["balance"] -= amount
        self.accounts[to_acc]["balance"] += amount

        self.accounts[from_acc]["history"].append(
            Transaction.create(-amount, "Przelew WYCHODZĄCY", f"Do: {to_acc}")
        )
        self.accounts[to_acc]["history"].append(
            Transaction.create(amount, "Przelew PRZYCHODZĄCY", f"Od: {from_acc}")
        )
        self._save_data()
        return True, "Przelew zrealizowany pomyślnie."


# === INTERFEJS UŻYTKOWNIKA ===
def get_float_input(prompt: str) -> float:
    while True:
        try:
            return float(input(prompt).replace(',', '.'))
        except ValueError:
            print("Błąd: Wprowadź poprawną liczbę (np. 150.50).")

def main():
    bank = Bank()

    while True:
        print("\n" + "="*30)
        print("    SYSTEM BANKOWY v1.0")
        print("="*30)
        print("1. Zaloguj się na konto")
        print("2. Załóż nowe konto")
        print("3. Wyjście")
        
        choice = input("Wybierz opcję (1-3): ")

        if choice == '3':
            print("Zamykanie systemu. Do widzenia!")
            break

        elif choice == '2':
            print("\n--- ZAKŁADANIE KONTA ---")
            acc_num = input("Podaj nowy numer konta (np. 1234): ")
            owner = input("Podaj imię i nazwisko właściciela: ")
            pin = input("Ustal 4-cyfrowy PIN: ")
            
            success, msg = bank.create_account(acc_num, owner, pin)
            print(f">>> {msg}")

        elif choice == '1':
            print("\n--- LOGOWANIE ---")
            acc_num = input("Numer konta: ")
            pin = input("PIN: ")
            
            account = bank.authenticate(acc_num, pin)
            if not account:
                print(">>> Błąd: Nieprawidłowy numer konta lub PIN.")
                continue

            # Pętla zalogowanego użytkownika
            while True:
                print(f"\n--- ZALOGOWANO: {account['owner']} (Konto: {acc_num}) ---")
                print(f"Bieżące saldo: {account['balance']:.2f} PLN")
                print("1. Wpłata gotówki")
                print("2. Wypłata gotówki")
                print("3. Przelew na inne konto")
                print("4. Historia transakcji")
                print("5. Wyloguj się")

                sub_choice = input("Wybierz akcję (1-5): ")

                if sub_choice == '5':
                    print("Wylogowano pomyślnie.")
                    break
                
                elif sub_choice == '1':
                    amount = get_float_input("Podaj kwotę do wpłaty: ")
                    success, msg = bank.deposit(acc_num, amount)
                    print(f">>> {msg}")

                elif sub_choice == '2':
                    amount = get_float_input("Podaj kwotę do wypłaty: ")
                    success, msg = bank.withdraw(acc_num, amount)
                    print(f">>> {msg}")

                elif sub_choice == '3':
                    to_acc = input("Podaj numer konta docelowego: ")
                    amount = get_float_input("Podaj kwotę przelewu: ")
                    success, msg = bank.transfer(acc_num, to_acc, amount)
                    print(f">>> {msg}")

                elif sub_choice == '4':
                    print("\n--- HISTORIA TRANSAKCJI ---")
                    if not account["history"]:
                        print("Brak operacji na koncie.")
                    else:
                        for t in account["history"]:
                            print(Transaction.format(t))

if __name__ == "__main__":
    main()

File: Bank.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Xml.Serialization;

namespace BankingSystem
{
    // Modele danych (muszą być publiczne dla serializacji XML)
    public class Transaction
    {
        public string Timestamp { get; set; }
        public decimal Amount { get; set; }
        public string Type { get; set; }
        public string Details { get; set; }

        public override string ToString()
        {
            return $"[{Timestamp}] {Type,-20} | {Amount,10:F2} PLN | {Details}";
        }
    }

    public class Account
    {
        public string AccountNumber { get; set; }
        public string Owner { get; set; }
        public string Pin { get; set; }
        public decimal Balance { get; set; }
        public List<Transaction> History { get; set; } = new List<Transaction>();
    }

    // Logika banku
    public class Bank
    {
        private const string DataFile = "bank_data.xml";
        private List<Account> Accounts;

        public Bank()
        {
            Accounts = LoadData();
        }

        private List<Account> LoadData()
        {
            if (File.Exists(DataFile))
            {
                try
                {
                    XmlSerializer serializer = new XmlSerializer(typeof(List<Account>));
                    using (FileStream fs = new FileStream(DataFile, FileMode.Open))
                    {
                        return (List<Account>)serializer.Deserialize(fs);
                    }
                }
                catch
                {
                    return new List<Account>();
                }
            }
            return new List<Account>();
        }

        private void SaveData()
        {
            try
            {
                XmlSerializer serializer = new XmlSerializer(typeof(List<Account>));
                using (StreamWriter writer = new StreamWriter(DataFile))
                {
                    serializer.Serialize(writer, Accounts);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Błąd zapisu pliku: " + ex.Message);
            }
        }

        private Account GetAccount(string accountNumber)
        {
            return Accounts.Find(a => a.AccountNumber == accountNumber);
        }

        public (bool Success, string Message) CreateAccount(string accountNumber, string owner, string pin)
        {
            if (GetAccount(accountNumber) != null)
                return (false, "Konto o tym numerze już istnieje.");

            Accounts.Add(new Account
            {
                AccountNumber = accountNumber,
                Owner = owner,
                Pin = pin,
                Balance = 0m
            });
            
            SaveData();
            return (true, "Konto zostało pomyślnie utworzone.");
        }

        public Account Authenticate(string accountNumber, string pin)
        {
            Account account = GetAccount(accountNumber);
            if (account != null && account.Pin == pin)
                return account;
            
            return null;
        }

        public (bool Success, string Message) Deposit(string accountNumber, decimal amount)
        {
            if (amount <= 0) return (false, "Kwota musi być większa niż zero.");

            Account account = GetAccount(accountNumber);
            account.Balance += amount;
            account.History.Add(new Transaction 
            { 
                Timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"), 
                Amount = amount, 
                Type = "Wpłata", 
                Details = "Wpłata własna" 
            });

            SaveData();
            return (true, "Wpłata zakończona sukcesem.");
        }

        public (bool Success, string Message) Withdraw(string accountNumber, decimal amount)
        {
            if (amount <= 0) return (false, "Kwota musi być większa niż zero.");

            Account account = GetAccount(accountNumber);
            if (account.Balance < amount) return (false, "Brak wystarczających środków.");

            account.Balance -= amount;
            account.History.Add(new Transaction 
            { 
                Timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"), 
                Amount = -amount, 
                Type = "Wypłata", 
                Details = "Wypłata w gotówce" 
            });

            SaveData();
            return (true, "Wypłata zakończona sukcesem.");
        }

        public (bool Success, string Message) Transfer(string fromAccount, string toAccount, decimal amount)
        {
            if (amount <= 0) return (false, "Kwota musi być większa niż zero.");
            if (fromAccount == toAccount) return (false, "Nie można wykonać przelewu na to samo konto.");
            
            Account sender = GetAccount(fromAccount);
            Account receiver = GetAccount(toAccount);

            if (receiver == null) return (false, "Konto docelowe nie istnieje.");
            if (sender.Balance < amount) return (false, "Brak wystarczających środków na koncie.");

            // Atomowa operacja
            sender.Balance -= amount;
            receiver.Balance += amount;

            string timeNow = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");

            sender.History.Add(new Transaction { Timestamp = timeNow, Amount = -amount, Type = "Przelew WYCH.", Details = $"Do: {toAccount}" });
            receiver.History.Add(new Transaction { Timestamp = timeNow, Amount = amount, Type = "Przelew PRZYCH.", Details = $"Od: {fromAccount}" });

            SaveData();
            return (true, "Przelew zrealizowany pomyślnie.");
        }
    }

    // Interfejs użytkownika
    class Program
    {
        static decimal GetDecimalInput(string prompt)
        {
            while (true)
            {
                Console.Write(prompt);
                string input = Console.ReadLine().Replace('.', ','); // Obsługa polskiego formatu liczb
                if (decimal.TryParse(input, out decimal result))
                {
                    return result;
                }
                Console.WriteLine("Błąd: Wprowadź poprawną liczbę (np. 150,50).");
            }
        }

        static void Main(string[] args)
        {
            Bank bank = new Bank();

            while (true)
            {
                Console.WriteLine("\n==============================");
                Console.WriteLine("    SYSTEM BANKOWY v1.0 (C#)");
                Console.WriteLine("==============================");
                Console.WriteLine("1. Zaloguj się na konto");
                Console.WriteLine("2. Załóż nowe konto");
                Console.WriteLine("3. Wyjście");
                Console.Write("Wybierz opcję (1-3): ");
                
                string choice = Console.ReadLine();

                if (choice == "3")
                {
                    Console.WriteLine("Zamykanie systemu. Do widzenia!");
                    break;
                }
                else if (choice == "2")
                {
                    Console.WriteLine("\n--- ZAKŁADANIE KONTA ---");
                    Console.Write("Podaj nowy numer konta (np. 1234): ");
                    string accNum = Console.ReadLine();
                    Console.Write("Podaj imię i nazwisko właściciela: ");
                    string owner = Console.ReadLine();
                    Console.Write("Ustal 4-cyfrowy PIN: ");
                    string pin = Console.ReadLine();

                    var result = bank.CreateAccount(accNum, owner, pin);
                    Console.WriteLine($"\n>>> {result.Message}");
                }
                else if (choice == "1")
                {
                    Console.WriteLine("\n--- LOGOWANIE ---");
                    Console.Write("Numer konta: ");
                    string accNum = Console.ReadLine();
                    Console.Write("PIN: ");
                    string pin = Console.ReadLine();

                    Account account = bank.Authenticate(accNum, pin);
                    if (account == null)
                    {
                        Console.WriteLine("\n>>> Błąd: Nieprawidłowy numer konta lub PIN.");
                        continue;
                    }

                    // Pętla zalogowanego użytkownika
                    while (true)
                    {
                        Console.WriteLine($"\n--- ZALOGOWANO: {account.Owner} (Konto: {accNum}) ---");
                        Console.WriteLine($"Bieżące saldo: {account.Balance:F2} PLN");
                        Console.WriteLine("1. Wpłata gotówki");
                        Console.WriteLine("2. Wypłata gotówki");
                        Console.WriteLine("3. Przelew na inne konto");
                        Console.WriteLine("4. Historia transakcji");
                        Console.WriteLine("5. Wyloguj się");
                        Console.Write("Wybierz akcję (1-5): ");

                        string subChoice = Console.ReadLine();

                        if (subChoice == "5")
                        {
                            Console.WriteLine("Wylogowano pomyślnie.");
                            break;
                        }
                        else if (subChoice == "1")
                        {
                            decimal amount = GetDecimalInput("Podaj kwotę do wpłaty: ");
                            var result = bank.Deposit(accNum, amount);
                            Console.WriteLine($"\n>>> {result.Message}");
                        }
                        else if (subChoice == "2")
                        {
                            decimal amount = GetDecimalInput("Podaj kwotę do wypłaty: ");
                            var result = bank.Withdraw(accNum, amount);
                            Console.WriteLine($"\n>>> {result.Message}");
                        }
                        else if (subChoice == "3")
                        {
                            Console.Write("Podaj numer konta docelowego: ");
                            string toAcc = Console.ReadLine();
                            decimal amount = GetDecimalInput("Podaj kwotę przelewu: ");
                            var result = bank.Transfer(accNum, toAcc, amount);
                            Console.WriteLine($"\n>>> {result.Message}");
                        }
                        else if (subChoice == "4")
                        {
                            Console.WriteLine("\n--- HISTORIA TRANSAKCJI ---");
                            if (account.History.Count == 0)
                            {
                                Console.WriteLine("Brak operacji na koncie.");
                            }
                            else
                            {
                                foreach (var t in account.History)
                                {
                                    Console.WriteLine(t);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
